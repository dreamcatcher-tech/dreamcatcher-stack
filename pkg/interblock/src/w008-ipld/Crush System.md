# The Crush System

This collection of classes are designed to break a Pulse down into IPFS compatible chunks, for efficient hashing.

## Requirements

1. use the inheritance of IpldInterface as identifying feature of classes that get crushed
1. crush relies on an unbroken chain of crushable classes from the root down
   1. ie: cannot have a pojo that holds a crush class in it and expect it to work
1. arrays of crush classes follow these rules
1. when flagged, some children may be 'punched out' as CID links, and reinflated back in
1. any tree of classes can be reinflated from ipfs back into classes
1. ipfs blocks representing the different between the previous 'crush' process can be fetched then persisted in ipfs storage
1. be able to directly represent an IPFS primitive as a class, such as a CID - these are needed to signal to the ORM that it should stop rehydration
1. Be able to connect a set of IPLD schemas to the classes, to check them for correctness on crush and uncrush
   1. This check should be disablable for speed
