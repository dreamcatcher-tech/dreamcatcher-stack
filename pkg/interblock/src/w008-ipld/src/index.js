export { Binary } from './Binary'
export { RequestId } from './RequestId'
export { Request } from './Request'
export { RxRequest } from './RxRequest'
export { Reply } from './Reply'
export { RxReply } from './RxReply'
export { Address } from './Address'
export { PulseLink, HistoricalPulseLink } from './PulseLink'
export { TxQueue } from './TxQueue'
export { Tx } from './Tx'
export { RxQueue } from './RxQueue'
export { Rx } from './Rx'
export { Channel } from './Channel'
export { Config } from './Config'
export { PublicKey } from './PublicKey'
export { Validators } from './Validators'
export { ChannelsHamt } from './ChannelsHamt'
export { ChildrenHamt } from './ChildrenHamt'
export { UplinksHamt } from './UplinksHamt'
export { DownlinksHamt } from './DownlinksHamt'
export { SymlinksHamt } from './SymlinksHamt'
export { AddressesHamt } from './AddressesHamt'
export { Channels } from './Channels'
export { Transmissions } from './Transmissions'
export { Network } from './Network'
export { AsyncRequest } from './AsyncRequest'
export { AsyncTrail } from './AsyncTrail'
export { Pending } from './Pending'
export { State } from './State'
export { Timestamp } from './Timestamp'
export { Dmz } from './Dmz'
export { Provenance } from './Provenance'
export { Interpulse } from './Interpulse'
export { Pulse } from './Pulse'
export { Keypair } from './Keypair'
export { decode } from './IpldUtils'
export { IpldInterface } from './IpldInterface'
export { Hamt } from './Hamt'
