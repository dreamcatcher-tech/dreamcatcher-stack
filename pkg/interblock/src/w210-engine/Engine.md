## Overview

The purpose of the Engine is to execute and coordinate all possible chain behaviours that are valid. Behaviours include: merging, forking, consensus conflicts, verification, multi chain interaction, versioning, covenant upgrades

The Engine has 3 types of event that it can process:

1. Interpulse updates - new interpulse for a chain we validate has occured
1. Pool updates - where there is a new pool for a chain
1. Pulse updates - a new Pulse as been formed

The Engine operates in local mode by default, meaning that it uses javascripts single threaded nature to guarantee locks. The engine can be connected to a hints service, which does the job of sharing the work of this engine with other engines, and can sometimes trigger the 3 engine events out of turn because they were caused externally. Interpulse updates trigger a pool update which triggers a pulse update. Pool updates trigger a Pulse update. Pulse updates trigger a pool update as the pool need to be reconciled with the new Pulse.

Pools are used to propose the next pulse, with a Pool update being used to send round something with your signature on it

## Pierce

Pierce is the means by which external Requests and Replies enter into chainland. Pierce is considered a form of pool update.

## Subscriptions

Clients will normally subscribe to Pulse updates. Validators will always subscribe to Interpulse updates, and in a multivalidator group, will also subscribe to Pool updates.

## TODO

Engines operate in one of four roles relative to a given chain:

1. proposing
2. validating
3. witnessing
4. consuming

## Effects

Execution secondarily deals with the running of side effects. This is because the loading of a covenant is the same for block making as it is for effects.
