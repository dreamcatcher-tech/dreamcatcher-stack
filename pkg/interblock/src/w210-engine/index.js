export { Engine } from './src/Engine'
