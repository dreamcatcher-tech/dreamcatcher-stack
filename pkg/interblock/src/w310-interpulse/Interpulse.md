## Ongoing Monitoring

The Engine controls active monitoring of what the latest block is, and it undertakes to resolve the hints it gets about there being new interpulses available for it to process.

This service is not started by default, must be switched on, and must also be shutdown to conclude testing.

If the service is not turned on, then the engine operates in local mode only. In local mode it can have some activity injected into it, and it responds to its own generated events and piercings, but that is all. The service is the network.
