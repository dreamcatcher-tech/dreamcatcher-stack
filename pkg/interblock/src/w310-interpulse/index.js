export * from './src/Interpulse'
export * from './src/Crisp'
export * from './src/Syncer'
