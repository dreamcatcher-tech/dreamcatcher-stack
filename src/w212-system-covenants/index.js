import { covenantIdModel } from '../w015-models'

import collection from './src/collection'
collection.covenantId = covenantIdModel.create('collection')

import datum from './src/datum'

import dpkg from './src/dpkg'
dpkg.covenantId = covenantIdModel.create('dpkg')

import shell from './src/shell'
shell.covenantId = covenantIdModel.create('shell')

import net from '../w020-emulators/src/netFactory'
net.covenantId = covenantIdModel.create('net')

import socket from '../w020-emulators/src/socketFactory'
socket.covenantId = covenantIdModel.create('socket')

import hyper from './src/hyper'
hyper.covenantId = covenantIdModel.create('hyper')

const unity = { reducer: (state = {}) => state }
unity.covenantId = covenantIdModel.create('unity')

import probe from './src/probe'
probe.covenantId = covenantIdModel.create('probe')

Object.values({
  collection,
  datum,
  dpkg,
  hyper,
  net,
  probe,
  shell,
  socket,
  unity,
}).forEach(Object.freeze)

export { collection, datum, dpkg, hyper, net, probe, shell, socket, unity }
