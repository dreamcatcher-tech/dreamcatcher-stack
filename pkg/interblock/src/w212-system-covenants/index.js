import * as collection from './src/collection.js'
import * as datum from './src/datum.js'
import * as net from './src/net.js'
import * as shell from './src/shell'
import * as hyper from './src/hyper'
import * as probe from './src/probe'
import * as unity from './src/unity'
import * as root from './src/unity'
import * as covenant from './src/covenant'
import * as ai from './src/ai.js'
import * as threads from './src/threads.js'
import * as goalie from './src/goalie.js'
import * as hal from './src/hal.js'
import * as bookMaker from './src/book-maker.js'
import * as stucks from './src/stucks.js'
import { Address } from '../w008-ipld/index.mjs'

export {
  collection,
  datum,
  hyper,
  probe,
  shell,
  unity,
  root,
  covenant,
  net,
  ai,
  threads,
  goalie,
  hal,
  bookMaker,
  stucks,
}

/**
 * Idea here is to provide a set of functions that can be used by any covenant to lock the current version of the system covenants that are in use.
 * A publisher might wish to reuse these classes to lock their own publications to a specific chain.
 * We can use techniques like 'pulselink + 1' to allow reference to the pulselink that we are contained within, whilst still using hashes.
 * importMap would need to override the explicit pulselink that was specified.
 */
const ADDRESS = Address.createCI('TODO create a chain for system covenants')
// const LATEST = PulseLink.generate() // TODO inject the latest pulselink
class System {
  constructor(name) {}
  getPulseLink() {
    // return the pulselink for the whole system chain, as recent as we know it
  }
  shell(version) {
    // allow to ask for a specific version ?
    return new System('shell')
  }
}
