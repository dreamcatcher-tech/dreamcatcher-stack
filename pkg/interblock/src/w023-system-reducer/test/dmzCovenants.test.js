import { Interpulse } from '../../w310-interpulse'
import { Request } from '../../w008-ipld'
import { Engine } from '../../w210-engine'
import { unity } from '../../w212-system-covenants'
import Debug from 'debug'
const debug = Debug('tests')

describe('dmzCovenants', () => {
  test('basic', async () => {
    const engine = await Interpulse.createCI()
    const request = Request.createGetCovenantState()
    const action = { ...request }
    const result = await engine.dispatch(action, '.')
    expect(result).toMatchSnapshot()
    debug(`result`, Object.keys(result))
    const latest = engine.selfLatest
    await engine.add('child1')
    const cr = await engine.dispatch(action, 'child1')
    expect(cr).toMatchSnapshot()
  })
  test('publish', async () => {
    const engine = await Interpulse.createCI()
    const covenant = { ...unity, name: 'published' }
    const { path } = await engine.publish('published', covenant)
    const latest = await engine.latest(path)
    expect(latest.provenance.dmz.covenant).toEqual('covenant')
    expect(latest.getState().toJS().name).toEqual('published')
    await engine.add('test', { covenant: path })
  })
})
