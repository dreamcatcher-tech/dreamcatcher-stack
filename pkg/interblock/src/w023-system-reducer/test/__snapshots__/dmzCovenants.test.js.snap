// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dmzCovenants basic 1`] = `
{
  "api": {
    "add": {
      "additionalProperties": false,
      "description": "Add a new chain at the optional path, with optional given installer.  If no path is given, a reasonable default will be chosen",
      "properties": {
        "installer": {
          "oneOf": [
            {
              "type": "string",
            },
            {
              "default": {},
              "type": "object",
            },
          ],
        },
        "path": {
          "type": "string",
        },
      },
      "required": [],
      "title": "ADD",
      "type": "object",
    },
    "cat": {
      "additionalProperties": false,
      "description": "Return the state as an object at the given path",
      "properties": {
        "path": {
          "default": ".",
          "type": "string",
        },
      },
      "required": [
        "path",
      ],
      "title": "CAT",
      "type": "object",
    },
    "cd": {
      "additionalProperties": false,
      "description": "Change directory to the given path",
      "properties": {
        "path": {
          "default": ".",
          "type": "string",
        },
      },
      "required": [
        "path",
      ],
      "title": "CD",
      "type": "object",
    },
    "covenant": {
      "additionalProperties": false,
      "description": "Return the state of a published covenant",
      "properties": {
        "path": {
          "default": ".",
          "type": "string",
        },
      },
      "required": [
        "path",
      ],
      "title": "COVENANT",
      "type": "object",
    },
    "dispatch": {
      "additionalProperties": false,
      "description": "Dispatch an action to a remote chain",
      "properties": {
        "action": {
          "type": "object",
        },
        "path": {
          "default": ".",
          "type": "string",
        },
      },
      "required": [
        "action",
        "path",
      ],
      "title": "DISPATCH",
      "type": "object",
    },
    "ln": {
      "description": "Link to target path.
        Linking is act of inserting one Object as the child of another
        which allows an Object to be the child of more than one parent.
        This operation is essential to application data structures
        as opposed to simple filesystem data structures, which are 
        usually a tree",
      "properties": {
        "linkName": {
          "description": "defaults to the target name.  Must not have any pathing",
          "type": "string",
        },
        "target": {
          "type": "string",
        },
      },
      "required": [
        "target",
      ],
      "title": "LN",
      "type": "object",
    },
    "login": {
      "additionalProperties": false,
      "description": "Authenticate with a remote app complex
  Loop the user through a signon process that links
  The current machine pubkey to their interblock user chain.
  When this occurs, the guest chain will transition to the
  user chain, and the prompt will change from "guest" to "user"
      ",
      "properties": {
        "chainId": {
          "type": "string",
        },
        "credentials": {
          "type": "object",
        },
      },
      "required": [
        "chainId",
        "credentials",
      ],
      "title": "LOGIN",
      "type": "object",
    },
    "ls": {
      "additionalProperties": false,
      "description": "List all children, and any actions available in the chain at the given path",
      "properties": {
        "path": {
          "default": ".",
          "type": "string",
        },
      },
      "required": [
        "path",
      ],
      "title": "LS",
      "type": "object",
    },
    "mount": {
      "additionalProperties": false,
      "description": "Attempt to mount the given chainId at the given mountPath.
      This will make an entry in mtab if there is not one already.",
      "properties": {
        "chainId": {
          "pattern": "Qm[1-9A-Za-z]{44}",
          "type": "string",
        },
        "name": {
          "type": "string",
        },
        "noSkip": {
          "description": "When fetching updates for the remote, ensure every pulse in the lineage is fetched, as opposed to lazily fetching only the latest known",
          "type": "boolean",
        },
      },
      "required": [
        "chainId",
        "name",
      ],
      "title": "MOUNT",
      "type": "object",
    },
    "multiaddr": {
      "description": "Add a new multiaddr to network memory",
      "properties": {
        "multiaddr": {
          "type": "string",
        },
      },
      "required": [
        "multiaddr",
      ],
      "title": "MULTIADDR",
      "type": "object",
    },
    "peer": {
      "description": "Add a new multiaddr to network memory",
      "properties": {
        "chainId": {
          "type": "string",
        },
        "peerId": {
          "type": "string",
        },
      },
      "required": [
        "peerId",
        "chainId",
      ],
      "title": "PEER",
      "type": "object",
    },
    "ping": {
      "additionalProperties": true,
      "description": "Ping a remote chain",
      "properties": {
        "message": {
          "type": "object",
        },
        "to": {
          "type": "string",
        },
      },
      "required": [],
      "title": "PING",
      "type": "object",
    },
    "publish": {
      "additionalProperties": false,
      "description": "Make a covenant ready for consumption",
      "properties": {
        "covenant": {
          "description": "The state of the pulished covenant chain",
          "type": "object",
        },
        "name": {
          "description": "A friendly hint for consumers",
          "type": "string",
        },
        "parentPath": {
          "default": ".",
          "description": "Path to the publication chain.  You must have permission to update this chain.  If the path does not exist but the parent does, a new default child will be created",
          "type": "string",
        },
      },
      "required": [
        "name",
        "covenant",
        "parentPath",
      ],
      "title": "PUBLISH",
      "type": "object",
    },
    "rm": {
      "additionalProperties": false,
      "description": "Attempt to remove the chain at the given path, and all its children",
      "properties": {
        "history": {
          "default": false,
          "description": "Remove the history too",
          "type": "boolean",
        },
        "path": {
          "default": ".",
          "type": "string",
        },
      },
      "required": [
        "path",
      ],
      "title": "RM",
      "type": "object",
    },
    "validators": {
      "description": "
      View, change the validator set of a chain or group of chains.
      Recursively change all validators of the chains children.
      Validators must accept the role before the handover is complete.
      Can be used to force a change if a chain has stalled.
      ",
      "title": "VALIDATORS",
      "type": "object",
    },
  },
  "installer": {
    "state": {
      "root": "/",
      "wd": "/",
    },
  },
  "name": "shell",
}
`;

exports[`dmzCovenants basic 2`] = `
{
  "api": {},
  "installer": {},
  "name": "unity",
}
`;
